service/history/hsm/tasks.go

```diff
@@ -85,3 +73,12 @@ type TaskSerializer interface {
 	Serialize(Task) ([]byte, error)
 	Deserialize(data []byte, attrs TaskAttributes) (Task, error)
 }
+
+// ValidateNotTransitioned returns a [consts.ErrStaleReference] if the machine has transitioned since the task was
+// generated.
+func ValidateNotTransitioned(ref *persistencespb.StateMachineRef, node *Node) error {
+	if ref.MachineTransitionCount != node.InternalRepr().TransitionCount {
```

Comment:
I thought we want to deprecate TransitionCount, no? so should use lastUpdateVersionedTransition if possible.

Reply:
I changed my mind here, and this is more correct since a sub state machine may transition multiple times in a single mutable state transition.

I might not even have this helper eventually, for now it's pretty convenient.